File: ts/src/grants.ts
/**
 * Grants (SPEC §6): signed, attenuable delegation chains.
 * A principal signs a root block granting a holder key some authority; any holder may
 * append a block delegating a narrower grant to another key.
 */
import { b64u, fromUtf8, unb64u, utf8 } from './b64.js';
import { canonical } from './canonical.js';
import { type KeyPair, keyPair, sha256, sign, verify } from './crypto.js';
import { fmtMoney } from './lens.js';
import type {
  ConsentRequest,
  Money,
  Proof,
  Proposal,
  Risk,
  Verb,
} from './types.js';

export interface Limit {
  max: number;
  currency: string;
}

export type Caveat =
  | { svc: string[] }
  | { verbs: Verb[] }
  | { can: string[] }
  | { exp: number }
  | { nbf: number }
  | { per: Limit }
  | { spend: Limit }
  | { risk: Risk }
  | { only: string };

export interface Block {
  p: Record<string, unknown> & { sub: string; caveats: Caveat[]; iat: number };
  s: string;
}

const PREFIX = 'pg1.';
const RISK_ORDER: Record<Risk, number> = { low: 0, medium: 1, high: 2 };
const now = () => Math.floor(Date.now() / 1000);

export function encodeGrant(blocks: Block[]): string {
  return PREFIX + b64u(utf8(canonical(blocks)));
}

export function decodeGrant(token: string): Block[] {
  if (!token.startsWith(PREFIX)) {
    throw new Error('not a pg1 grant');
  }

  const blocks = JSON.parse(fromUtf8(unb64u(token.slice(PREFIX.length))));

  if (!Array.isArray(blocks) || blocks.length === 0) {
    throw new Error('grant has no blocks');
  }

  for (const b of blocks) {
    if (
      typeof b?.s !== 'string' ||
      typeof b?.p?.sub !== 'string' ||
      !Array.isArray(b?.p?.caveats)
    ) {
      throw new Error('malformed block');
    }
  }

  return blocks;
}

export const blockId = (b: Block) => sha256(b.s);

/** Issue a root grant from `principal` to the key `to`. */
export async function issueGrant(opts: {
  principal: KeyPair | string;
  to: string;
  caveats?: Caveat[];
  iat?: number;
  nonce?: string;
}): Promise<string> {
  const principal =
    typeof opts.principal === 'string'
      ? await keyPair(opts.principal)
      : opts.principal;
  const p = {
    iss: principal.public,
    sub: opts.to,
    caveats: opts.caveats ?? [],
    iat: opts.iat ?? now(),
    nonce:
      opts.nonce ?? b64u(globalThis.crypto.getRandomValues(new Uint8Array(12))),
  };

  return encodeGrant([{ p, s: await sign(principal.seed, canonical(p)) }]);
}

/** Attenuate: the current holder (by seed) delegates a narrower grant to `to`. */
export async function delegateGrant(
  token: string,
  opts: {
    holder: KeyPair | string;
    to: string;
    caveats?: Caveat[];
    iat?: number;
  },
): Promise<string> {
  const blocks = decodeGrant(token);
  const holder =
    typeof opts.holder === 'string' ? await keyPair(opts.holder) : opts.holder;
  const last = blocks[blocks.length - 1];

  if (last.p.sub !== holder.public) {
    throw new Error('only the current holder can delegate this grant');
  }

  const p = {
    prev: await sha256(last.s),
    sub: opts.to,
    caveats: opts.caveats ?? [],
    iat: opts.iat ?? now(),
  };

  return encodeGrant([
    ...blocks,
    { p, s: await sign(holder.seed, canonical(p)) },
  ]);
}

/**
 * Consent grant (SPEC §6.6): one-shot approval of one exact proposal. It is scoped to COMMIT
 * of that proposal's capability at that service, so it authorizes nothing else.
 */
export function consentGrant(opts: {
  principal: KeyPair | string;
  agent: string;
  consent: Pick<ConsentRequest, 'service' | 'capability' | 'hash' | 'expires'>;
}): Promise<string> {
  const c = opts.consent;

  return issueGrant({
    principal: opts.principal,
    to: opts.agent,
    caveats: [
      { svc: [c.service] },
      { verbs: ['COMMIT'] },
      { can: [c.capability] },
      { only: c.hash },
      { exp: c.expires },
    ],
  });
}

/**
 * A consent request packed for a human to approve out of band (`parley approve <code>`).
 * `detail` carries the full proposal (as the agent saw it) so the approver can show its
 * effects and re-check the hash, instead of trusting a service-written summary.
 */
export function consentCode(
  c: ConsentRequest,
  detail?: Omit<Proposal, 'data'>,
): string {
  const { data: _d, ...d } = (detail ?? {}) as Proposal;

  return `pc1.${b64u(utf8(canonical(detail ? { ...c, detail: d } : c)))}`;
}

export function decodeConsentCode(
  code: string,
): ConsentRequest & { detail?: Proposal } {
  if (!code.startsWith('pc1.')) {
    throw new Error('not a consent code (expected pc1.…)');
  }

  const c = JSON.parse(fromUtf8(unb64u(code.slice(4))));

  for (const k of [
    'proposal',
    'hash',
    'service',
    'capability',
    'principal',
    'summary',
  ]) {
    if (typeof c[k] !== 'string') {
      throw new Error(`consent code missing ${k}`);
    }
  }

  if (!Number.isSafeInteger(c.expires)) {
    throw new Error('consent code missing expires');
  }

  return c;
}

export interface GrantInfo {
  id: string;
  iss: string;
  holder: string;
  blocks: { id: string; sub: string; caveats: Caveat[]; iat: number }[];
}

export async function inspectGrant(token: string): Promise<GrantInfo> {
  const blocks = decodeGrant(token);
  const ids = await Promise.all(blocks.map(blockId));

  return {
    id: ids[0],
    iss: blocks[0].p.iss as string,
    holder: blocks[blocks.length - 1].p.sub,
    blocks: blocks.map((b, i) => ({
      id: ids[i],
      sub: b.p.sub,
      caveats: b.p.caveats,
      iat: b.p.iat,
    })),
  };
}

export interface CheckContext {
  service: string;
  verb: Verb;
  capability: string;
  now?: number;
  /** For COMMIT: the proposal being committed. */
  proposal?: { hash: string; cost: Money | null; risk: Risk };
  /** Principals the service trusts. */
  trusted: string[] | ((iss: string) => boolean);
  /** The key that signed the request proof; must equal the grant holder. */
  proofKey: string;
  /** Spend already attributed to a block id (for `spend` caveats). */
  spent?: (blockId: string) => number;
}

export type GrantCheck =
  | {
      ok: true;
      id: string;
      iss: string;
      holder: string;
      spendBlocks: { id: string; max: number }[];
    }
  | {
      ok: false;
      code: 'unauthorized' | 'forbidden' | 'consent_required';
      reason: string;
      iss?: string;
      need?: Caveat[];
    };

const CONSENTABLE = new Set(['per', 'spend', 'risk']);

const strList = (v: unknown) =>
  Array.isArray(v) && v.every((x) => typeof x === 'string');
const isLimit = (v: unknown) =>
  !!v &&
  typeof v === 'object' &&
  Number.isSafeInteger((v as Partial<Limit>).max) &&
  typeof (v as Partial<Limit>).currency === 'string';

/** Shape validators for each known caveat's value. */
const VALID_CAVEAT: Record<string, (v: unknown) => boolean> = {
  svc: strList,
  verbs: strList,
  can: strList,
  exp: Number.isSafeInteger,
  nbf: Number.isSafeInteger,
  per: isLimit,
  spend: isLimit,
  risk: (v) => typeof v === 'string' && Object.hasOwn(RISK_ORDER, v),
  only: (v) => typeof v === 'string',
};

/** Caveats with malformed values fail closed (as a hard failure). */
function malformed(k: string, v: unknown): string | null {
  // unknown keys are handled by caveatDenial
  const ok = Object.hasOwn(VALID_CAVEAT, k) ? VALID_CAVEAT[k](v) : true;

  return ok ? null : `malformed caveat ${JSON.stringify({ [k]: v })}`;
}

export function matchCapability(pattern: string, cap: string): boolean {
  return (
    pattern === '*' ||
    pattern === cap ||
    (pattern.endsWith('*') && cap.startsWith(pattern.slice(0, -1)))
  );
}

/** Verify a grant token against a request (SPEC §6.4). Never throws. */
export async function checkGrant(
  token: string,
  ctx: CheckContext,
): Promise<GrantCheck> {
  try {
    return await checkGrantUnsafe(token, ctx);
  } catch (e) {
    return {
      ok: false,
      code: 'forbidden',
      reason: `malformed grant content: ${(e as Error).message}`,
    };
  }
}

type GrantFailure = Extract<GrantCheck, { ok: false }>;

const unauthorized = (reason: string, iss?: string): GrantFailure => ({
  ok: false,
  code: 'unauthorized',
  reason,
  ...(iss === undefined ? {} : { iss }),
});

/** Check each block is chained to the previous one and signed by its holder; yields the final holder. */
async function verifyChain(
  blocks: Block[],
  iss: string,
): Promise<{ ok: true; holder: string } | GrantFailure> {
  let signer = iss;

  for (let i = 0; i < blocks.length; i++) {
    const b = blocks[i];

    if (i > 0 && b.p.prev !== (await sha256(blocks[i - 1].s))) {
      return unauthorized(`block ${i} is not chained to block ${i - 1}`);
    }

    if (!(await verify(signer, canonical(b.p), b.s))) {
      return unauthorized(`bad signature on block ${i}`);
    }

    signer = b.p.sub;
  }

  return { ok: true, holder: signer };
}

/** What a caveat is checked against: the request, the time, and the block it sits in. */
interface CaveatEnv {
  ctx: CheckContext;
  t: number;
  /** The proposal, for COMMIT only. */
  p: CheckContext['proposal'];
  blockId: string;
}

/**
 * Per-key checks of a well-formed caveat value (already shape-validated by `malformed`):
 * why it denies the request, or null if it allows it.
 */
const CAVEAT_CHECKS: Record<
  string,
  (v: unknown, env: CaveatEnv) => string | null
> = {
  svc: (v, { ctx }) =>
    (v as string[]).includes(ctx.service)
      ? null
      : `not valid for service ${ctx.service}`,
  verbs: (v, { ctx }) =>
    (v as Verb[]).includes(ctx.verb) ? null : `does not allow ${ctx.verb}`,
  can: (v, { ctx }) =>
    (v as string[]).some((pat) => matchCapability(pat, ctx.capability))
      ? null
      : `does not cover ${ctx.capability}`,
  exp: (v, { t }) => (t < (v as number) ? null : 'grant has expired'),
  nbf: (v, { t }) => (t >= (v as number) ? null : 'grant is not valid yet'),
  per: (v, { p }) => {
    const limit = v as Limit;

    return p?.cost &&
      (p.cost.currency !== limit.currency || p.cost.amount > limit.max)
      ? `cost exceeds the per-commit limit of ${fmtMoney({ amount: limit.max, currency: limit.currency })}`
      : null;
  },
  spend: (v, { ctx, p, blockId }) => {
    const limit = v as Limit;

    if (!p?.cost) {
      return null;
    }

    const total = (ctx.spent?.(blockId) ?? 0) + p.cost.amount;

    return p.cost.currency !== limit.currency || total > limit.max
      ? `would exceed the spend limit of ${fmtMoney({ amount: limit.max, currency: limit.currency })}`
      : null;
  },
  risk: (v, { p }) =>
    p && RISK_ORDER[p.risk] > RISK_ORDER[v as Risk]
      ? `risk ${p.risk} exceeds ceiling ${v}`
      : null,
  only: (v, { ctx, p }) =>
    ctx.verb === 'COMMIT' && p?.hash !== v
      ? 'grant is bound to a different proposal'
      : null,
};

/** Why the single-key caveat `c` = `{k: v}` denies the request; unknown caveats always do. */
function caveatDenial(
  c: object,
  k: string,
  v: unknown,
  env: CaveatEnv,
): string | null {
  if (!Object.hasOwn(CAVEAT_CHECKS, k)) {
    return `unknown caveat ${JSON.stringify(c)}`;
  }

  return CAVEAT_CHECKS[k](v, env);
}

interface CaveatFailure {
  c: Caveat;
  why: string;
}

interface CaveatResults {
  /** Denials no one can override. */
  hard: CaveatFailure[];
  /** Denials a principal may approve (limits and risk ceilings). */
  soft: CaveatFailure[];
  /** `spend` blocks this COMMIT counts against. */
  spendBlocks: { id: string; max: number }[];
}

/** Check one caveat, recording a denial (hard or soft) or the spend block it charges. */
function evaluateCaveat(c: unknown, env: CaveatEnv, out: CaveatResults) {
  if (!c || typeof c !== 'object' || Array.isArray(c)) {
    out.hard.push({
      c: c as Caveat,
      why: `malformed caveat ${JSON.stringify(c)}`,
    });

    return;
  }

  const keys = Object.keys(c);
  const k = keys.length === 1 ? keys[0] : '';
  const v = (c as Record<string, unknown>)[k];
  const why = malformed(k, v) ?? caveatDenial(c, k, v, env);

  if (why) {
    (CONSENTABLE.has(k) && !why.startsWith('malformed')
      ? out.soft
      : out.hard
    ).push({ c: c as Caveat, why });
  } else if (k === 'spend' && env.ctx.verb === 'COMMIT') {
    out.spendBlocks.push({ id: env.blockId, max: (v as Limit).max });
  }
}

/** Check every caveat of every block. Fails closed: anything unrecognized is a hard denial. */
async function evaluateCaveats(
  blocks: Block[],
  ctx: CheckContext,
): Promise<CaveatResults> {
  const t = ctx.now ?? now();
  const p = ctx.verb === 'COMMIT' ? ctx.proposal : undefined;
  const out: CaveatResults = { hard: [], soft: [], spendBlocks: [] };

  for (const b of blocks) {
    const env = { ctx, t, p, blockId: await sha256(b.s) };
    const caveats: unknown[] = b.p.caveats; // decoded but not yet validated

    for (const c of caveats) {
      evaluateCaveat(c, env, out);
    }
  }

  return out;
}

async function checkGrantUnsafe(
  token: string,
  ctx: CheckContext,
): Promise<GrantCheck> {
  let blocks: Block[];

  try {
    blocks = decodeGrant(token);
  } catch (e) {
    return unauthorized(`malformed grant: ${(e as Error).message}`);
  }

  const iss = blocks[0].p.iss;

  if (typeof iss !== 'string') {
    return unauthorized('root block has no iss');
  }

  const chain = await verifyChain(blocks, iss);

  if (!chain.ok) {
    return chain;
  }

  const trusted =
    typeof ctx.trusted === 'function'
      ? ctx.trusted(iss)
      : ctx.trusted.includes(iss);

  if (!trusted) {
    return unauthorized(
      'grant is issued by a principal this service does not trust',
      iss,
    );
  }

  if (chain.holder !== ctx.proofKey) {
    return unauthorized('proof key is not the grant holder', iss);
  }

  const { hard, soft, spendBlocks } = await evaluateCaveats(blocks, ctx);

  if (hard.length) {
    return {
      ok: false,
      code: 'forbidden',
      reason: hard.map((h) => h.why).join('; '),
      iss,
      need: hard.map((h) => h.c),
    };
  }

  if (soft.length) {
    return {
      ok: false,
      code: 'consent_required',
      reason: soft.map((h) => h.why).join('; '),
      iss,
    };
  }

  return {
    ok: true,
    id: await sha256(blocks[0].s),
    iss,
    holder: chain.holder,
    spendBlocks,
  };
}

export interface ProofTarget {
  aud: string;
  verb: Verb;
  target: string;
}

export async function makeProof(
  seed: string,
  t: ProofTarget,
  ts = now(),
): Promise<Proof> {
  const kp = await keyPair(seed);

  return {
    key: kp.public,
    ts,
    sig: await sign(
      seed,
      canonical({ aud: t.aud, verb: t.verb, target: t.target, ts }),
    ),
  };
}

export async function checkProof(
  proof: Proof | undefined,
  t: ProofTarget,
  at = now(),
): Promise<string | null> {
  if (
    !proof ||
    typeof proof.key !== 'string' ||
    typeof proof.sig !== 'string' ||
    !Number.isSafeInteger(proof.ts)
  ) {
    return 'missing or malformed proof';
  }

  if (Math.abs(at - proof.ts) > 300) {
    return 'proof timestamp is outside the 300s window';
  }

  const ok = await verify(
    proof.key,
    canonical({ aud: t.aud, verb: t.verb, target: t.target, ts: proof.ts }),
    proof.sig,
  );

  return ok ? null : 'proof signature is invalid';
}

File: python/src/parley/grants.py
"""Grants: attenuable Ed25519 delegation chains (SPEC §6.2–6.4)."""

from __future__ import annotations

import os
import time
from collections.abc import Callable, Iterable, Mapping
from dataclasses import dataclass, field
from typing import Any

from ._json import CanonicalError, b64url_decode, b64url_encode, canonical_bytes, compact, loads, sha256_b64url
from .keys import KeyPair, parse_public_key, verify

TOKEN_PREFIX = "pg1."
RISK_ORDER = {"low": 0, "medium": 1, "high": 2}
# Caveats that, when they are the only ones failing a COMMIT, mean "ask the human" (§6.6).
CONSENT_CAVEATS = frozenset({"risk", "per", "spend"})
COMMIT_ONLY = frozenset({"per", "spend", "risk", "only"})

Trusted = Iterable[str] | Callable[[str], bool]


def block_id(block: Mapping[str, Any]) -> str:
    """``b64url(sha256(utf8(block.s)))``."""
    return sha256_b64url(block["s"].encode("utf-8"))


@dataclass(frozen=True)
class Grant:
    blocks: tuple[dict, ...]

    @property
    def id(self) -> str:
        return block_id(self.blocks[0])

    @property
    def block_ids(self) -> list[str]:
        return [block_id(b) for b in self.blocks]

    @property
    def principal(self) -> str:
        return self.blocks[0]["p"]["iss"]

    @property
    def holder(self) -> str:
        return self.blocks[-1]["p"]["sub"]

    def encode(self) -> str:
        return TOKEN_PREFIX + b64url_encode(canonical_bytes(list(self.blocks)))

    def delegate(self, holder: KeyPair, sub: str, caveats: list[dict], iat: int | None = None) -> Grant:
        """Append a block handing a narrower grant to ``sub``. ``holder`` must be the current holder."""
        if holder.public != self.holder:
            raise ValueError("only the grant's holder can delegate it")
        payload = {
            "prev": block_id(self.blocks[-1]),
            "sub": sub,
            "caveats": list(caveats),
            "iat": int(time.time()) if iat is None else iat,
        }
        return Grant(self.blocks + ({"p": payload, "s": holder.sign(canonical_bytes(payload))},))

    def __str__(self) -> str:
        return self.encode()


def issue_grant(
    issuer: KeyPair, sub: str, caveats: list[dict], iat: int | None = None, nonce: str | None = None
) -> Grant:
    """Issue a root grant from principal ``issuer`` to holder key ``sub``."""
    payload = {
        "iss": issuer.public,
        "sub": sub,
        "caveats": list(caveats),
        "iat": int(time.time()) if iat is None else iat,
        "nonce": b64url_encode(os.urandom(12)) if nonce is None else nonce,
    }
    return Grant(({"p": payload, "s": issuer.sign(canonical_bytes(payload))},))


def delegate_grant(grant: Grant | str, holder: KeyPair, sub: str, caveats: list[dict], iat: int | None = None) -> Grant:
    g = decode_grant(grant) if isinstance(grant, str) else grant
    return g.delegate(holder, sub, caveats, iat)


def consent_grant(principal: KeyPair, agent_key: str, consent: Mapping[str, Any]) -> Grant:
    """The one-shot grant a principal signs to approve a ``consent_required`` proposal (§6.6)."""
    return issue_grant(principal, agent_key, [{"only": consent["hash"]}, {"exp": consent["expires"]}])


def _is_int(v: Any) -> bool:
    return type(v) is int


def decode_grant(token: str) -> Grant:
    """Decode and structurally validate a token. Does not check signatures. Raises ValueError."""
    if not isinstance(token, str) or not token.startswith(TOKEN_PREFIX):
        raise ValueError("grant tokens start with pg1.")
    try:
        blocks = loads(b64url_decode(token[len(TOKEN_PREFIX):]).decode("utf-8"))
    except (ValueError, UnicodeDecodeError) as e:
        raise ValueError(f"grant token is not valid b64url JSON: {e}") from None
    if not isinstance(blocks, list) or not blocks:
        raise ValueError("a grant is a non-empty list of blocks")
    for i, b in enumerate(blocks):
        if not isinstance(b, dict) or not isinstance(b.get("p"), dict) or not isinstance(b.get("s"), str):
            raise ValueError(f"block {i} must be {{p, s}}")
        p = b["p"]
        need = ("iss", "sub", "nonce") if i == 0 else ("prev", "sub")
        if not all(isinstance(p.get(k), str) for k in need):
            raise ValueError(f"block {i} payload is missing {', '.join(need)}")
        if not isinstance(p.get("caveats"), list) or not _is_int(p.get("iat")):
            raise ValueError(f"block {i} payload needs caveats[] and an integer iat")
    return Grant(tuple(blocks))


@dataclass
class GrantContext:
    """What a request asks for, evaluated against caveats."""

    service: str
    verb: str
    capability: str | None
    now: int
    proposal: Mapping[str, Any] | None = None  # {hash, cost, risk}; COMMIT only
    spent: Mapping[str, int] = field(default_factory=dict)  # block id -> committed spend

    @classmethod
    def from_dict(cls, d: Mapping[str, Any]) -> GrantContext:
        return cls(d["service"], d["verb"], d.get("capability"), d["now"], d.get("proposal"), d.get("spent") or {})


@dataclass
class Verification:
    ok: bool
    code: str | None = None  # unauthorized | forbidden | consent_required
    message: str = ""
    grant: Grant | None = None
    failed: list[dict] = field(default_factory=list)  # caveats that were not satisfied

    @property
    def need(self) -> list[dict] | None:
        """For ``forbidden``: the caveats a grant would need to drop (SPEC §7 ``need``)."""
        return self.failed if self.code == "forbidden" else None

    @property
    def principal(self) -> str | None:
        return self.grant.principal if self.grant else None

    def spend_blocks(self) -> list[tuple[str, dict]]:
        """(block id, spend caveat) for every block carrying a ``spend`` caveat."""
        out = []
        if self.grant:
            for b in self.grant.blocks:
                for c in b["p"]["caveats"]:
                    if isinstance(c, dict) and "spend" in c:
                        out.append((block_id(b), c["spend"]))
        return out


def _matches(pattern: Any, name: str) -> bool:
    if not isinstance(pattern, str):
        return False
    if pattern.endswith("*"):
        return name.startswith(pattern[:-1])
    return pattern == name


def _money_ok(limit: Any, cost: Any, already: int = 0) -> bool:
    if not isinstance(limit, dict) or not _is_int(limit.get("max")) or not isinstance(limit.get("currency"), str):
        return False
    if cost is None:
        return True
    if not isinstance(cost, dict) or not _is_int(cost.get("amount")) or cost.get("currency") != limit["currency"]:
        return False
    return already + cost["amount"] <= limit["max"]


def _safe_int(v: Any) -> bool:
    return _is_int(v) and abs(v) <= 2**53 - 1


def _str_list(v: Any) -> bool:
    return isinstance(v, list) and all(isinstance(x, str) for x in v)


def _limit(v: Any) -> bool:
    return isinstance(v, dict) and _safe_int(v.get("max")) and isinstance(v.get("currency"), str)


_WELL_FORMED = {
    "svc": _str_list, "verbs": _str_list, "can": _str_list,
    "exp": _safe_int, "nbf": _safe_int,
    "per": _limit, "spend": _limit,
    "risk": lambda v: isinstance(v, str) and v in RISK_ORDER,
    "only": lambda v: isinstance(v, str),
}


def well_formed(caveat: Any) -> bool:
    """A known caveat name with a value of the right shape (SPEC §6.3)."""
    if not isinstance(caveat, dict) or len(caveat) != 1:
        return False
    (name, arg), = caveat.items()
    check = _WELL_FORMED.get(name)
    return check is not None and check(arg)


def check_caveat(caveat: Any, bid: str, ctx: GrantContext) -> bool:
    """True iff one caveat is satisfied. Unknown or malformed caveats fail closed."""
    if not well_formed(caveat):
        return False
    (name, arg), = caveat.items()
    if name in COMMIT_ONLY and ctx.verb != "COMMIT":
        return True  # known, and ignored outside COMMIT (§6.3)
    if name == "svc":
        return isinstance(arg, list) and ctx.service in arg
    if name == "verbs":
        return isinstance(arg, list) and ctx.verb in arg
    if name == "can":
        return isinstance(arg, list) and ctx.capability is not None and any(_matches(p, ctx.capability) for p in arg)
    if name == "exp":
        return _is_int(arg) and ctx.now < arg
    if name == "nbf":
        return _is_int(arg) and ctx.now >= arg
    prop = ctx.proposal or {}
    if name == "per":
        return _money_ok(arg, prop.get("cost"))
    if name == "spend":
        return _money_ok(arg, prop.get("cost"), int(ctx.spent.get(bid, 0)))
    if name == "risk":
        return arg in RISK_ORDER and prop.get("risk") in RISK_ORDER and RISK_ORDER[prop["risk"]] <= RISK_ORDER[arg]
    if name == "only":
        return isinstance(arg, str) and prop.get("hash") == arg
    return False


def _trusts(trusted: Trusted, key: str) -> bool:
    return trusted(key) if callable(trusted) else key in set(trusted)


def verify_grant(
    token: str | Grant, trusted: Trusted, proof_key: str, ctx: GrantContext | Mapping[str, Any]
) -> Verification:
    """Verify one grant for one request (SPEC §6.4). Never raises."""
    if not isinstance(ctx, GrantContext):
        ctx = GrantContext.from_dict(ctx)
    try:
        g = token if isinstance(token, Grant) else decode_grant(token)
    except ValueError as e:
        return Verification(False, "unauthorized", str(e))

    for i, b in enumerate(g.blocks):
        p = b["p"]
        signer = p["iss"] if i == 0 else g.blocks[i - 1]["p"]["sub"]
        if i > 0 and p["prev"] != block_id(g.blocks[i - 1]):
            return Verification(False, "unauthorized", f"block {i} does not chain to block {i - 1}", g)
        try:
            parse_public_key(p["sub"])
            msg = canonical_bytes(p)
        except (ValueError, CanonicalError) as e:
            return Verification(False, "unauthorized", f"block {i} is malformed: {e}", g)
        if not verify(signer, msg, b["s"]):
            return Verification(False, "unauthorized", f"block {i} signature does not verify", g)

    if not _trusts(trusted, g.principal):
        return Verification(False, "unauthorized", "the grant's principal is not trusted by this service", g)
    if g.holder != proof_key:
        return Verification(False, "unauthorized", "the proof key is not the grant's holder", g)

    failed = [c for b in g.blocks for c in b["p"]["caveats"] if not check_caveat(c, block_id(b), ctx)]
    if not failed:
        return Verification(True, grant=g)
    hard = [c for c in failed if _caveat_name(c) not in CONSENT_CAVEATS or not well_formed(c)]
    if hard:
        shown = ", ".join(compact(c) for c in hard[:3])
        return Verification(False, "forbidden", f"the grant does not allow this request ({shown})", g, hard)
    shown = ", ".join(compact(c) for c in failed[:3])
    return Verification(False, "consent_required", f"the proposal exceeds the grant's limits ({shown})", g, failed)


def _caveat_name(c: Any) -> str | None:
    return next(iter(c)) if isinstance(c, dict) and len(c) == 1 else None
